### PR TITLE
Hostile mob, spiders and carps update

### PR DESCRIPTION
**Pull request by Sindorman**
Summary:

- Hostile mobs now look for living mobs first, before looking other things(mechs and bots)

- Carps and giant spiders now attack shields if they see them. Notes: shield segment max strength is 10, and carp attacks 1-2 per attack.

- If carp is targeting mob, and there is a shield in the way to mob, they will destroy shield while keep pursuing mob.

- better, more generic code for targets

### CODE REVIEW

**Comment by Sindorman** [comment]
!review

File: code/modules/mob/living/simple_animal/hostile/carp.dm [hunk 1]
```diff
@@ -45,17 +45,137 @@
...
+	if(get_dist(src, target_mob) <= 1)	//Attacking
+		AttackingTarget()
+		attacked_times += 1
+		return 1
 
 /mob/living/simple_animal/hostile/carp/AttackingTarget()
-	. =..()
-	var/mob/living/L = .
-	if(istype(L))
+	setClickCooldown(attack_delay)
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/carp.dm]: Why is this necessary?

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 2]
```diff
@@ -28,47 +28,43 @@
...
+			if(F)
+				T = F
 				break
 
-		if(istype(A, /obj/machinery/bot))
-			var/obj/machinery/bot/B = A
-			if (B.health > 0)
-				T = B
-				break
+			else if(istype(A, /obj/mecha)) // Our line of sight stuff was already done in ListTargets().
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: So, here's a thought. Generalize this. At the moment, you have relatively meaningless overrides which simply add the functionality in the form of:
````dm
if (istype(A, /some/path) && SomeValidity(A))
    T = A
    break
````

Which is a lot of pointless reinventing of the wheel. Instead, this could be generalized as:
````c
var/list/target_type_validator_map = list(/some/type = CALLBACK(src, /proc/someFunction), /other/path = TRUE) // this would be a class variable

// ...

for (var/atom/A in targets)
    if (target_type_validator_map[A.type])
        var/datum/callback/cb = target_type_validator_map[A.type]
        if (!istype(cb))
            T = A // if it's not a callback, then we don't have a validator, ergo, always prefer this target!
        else if (cb.Invoke(A))
            T = A // Validator returned TRUE, valid target, go nuts.
        break
````

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 3]
```diff
@@ -28,47 +28,43 @@
...
 		return null
 
 	var/atom/T = null
 	var/lowest_health = INFINITY // Max you can get
 	stop_automated_movement = 0
-
-	for(var/atom/A in targets)
-
-		if(A == src)
+	for(var/mob/living/L in targets)
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: I'd like to see this loop removed. You're iterating the `targets` list twice. You can just prioritize livings in the main loop and break immediately as you find one.

File: code/modules/mob/living/simple_animal/hostile/carp.dm [hunk 4]
```diff
@@ -45,17 +45,137 @@
...
+		if((L.faction == src.faction) && !attack_same)
+			continue
+		if(L in friends)
+			continue
+
+		if(!L.stat && (L.health < lowest_health))
+			lowest_health = L.health
+			T = L
+	if(isnull(T))
+		for(var/atom/A in targets)
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/carp.dm]: See my comment about generalization. You could basically _almost_ remove this function override with that.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 5]
```diff
@@ -32,36 +40,19 @@
...
-			TL = A
-
-		else if(istype(A, /obj/mecha)) // Our line of sight stuff was already done in ListTargets().
-			var/obj/mecha/M = A
-			if (M.occupant)
-				T = M
+		if (target_type_validator_map[A.type])
+			var/datum/callback/cb = target_type_validator_map[A.type]
+			if (!istype(cb))
+				T = A // if it's not a callback, then we don't have a validator, ergo, always prefer this target!
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: Missing break from here, probably.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 6]
```diff
@@ -20,6 +20,14 @@
 	var/enroute = 0
 	var/list/targets = list()
 	var/attacked_times = 0
+	var/list/target_type_validator_map = list()
+	var/lowest_health = INFINITY // Max you can get
+
+/mob/living/bot/secbot/Initialize()
+	. = ..()
+	target_type_validator_map.add(/mob/living = CALLBACK(src, ./proc/living), /other/path = TRUE)
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: ooookay, that's not how you add stuff to a map. This should be:
```dm
	target_type_validator_map[/mob/living] = CALLBACK(src, ./proc/living)
	target_type_validator_map[/obj/mecha] = CALLBACK(src, ./proc/mecha)
	target_type_validator_map[/obj/machinery/bot] = CALLBACK(src, ./proc/bot)
```

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 7]
```diff
@@ -32,36 +40,19 @@
...
-				lowest_health = L.health
-				T = L
-		else if (isliving(A))
-			TL = A
-
-		else if(istype(A, /obj/mecha)) // Our line of sight stuff was already done in ListTargets().
-			var/obj/mecha/M = A
-			if (M.occupant)
-				T = M
+		if (target_type_validator_map[A.type])
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: Also I just realized, this doesn't cover subtypes. Which would be a logic error. Might be worth using this logic instead:
```dm
		if(A == src)
			continue

		var/datum/callback/cb = null
		for (var/type in target_type_validator_map)
			if (istype(A, type)
				cb= target_type_validator_map[type]

		if (!cb)
			continue
		else if (!istype(cb) || cb.Invoke(A))
			T = A
			break
```

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 8]
```diff
@@ -293,3 +284,29 @@ mob/living/simple_animal/hostile/hitby(atom/movable/AM as mob|obj,var/speed = TH
 		spawn(10)
 			if(!src.stat)
 				horde()
+
+/mob/living/simple_animal/hostile/proc/living(var/atom/A)
+	if(isliving(A))
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: These type checks should be unnecessary. You can just declare it as the wanted type in the parameter.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm (hunk 6 above)
Comment (Sindorman) [reply code/modules/mob/living/simple_animal/hostile/hostile.dm]: I was confused by your comment before.

**Comment by skull132** [comment]
Pushed over one update, due to not meeting deadline.

File: code/modules/mob/living/simple_animal/hostile/carp.dm (hunk 1 above)
Comment (Sindorman) [reply code/modules/mob/living/simple_animal/hostile/carp.dm]: That is what parent has, since it is the override

**Comment by Sindorman** [comment]
@skull132 It is done.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 9]
```diff
@@ -28,47 +40,25 @@
...
-	for(var/atom/A in targets)
-
+	T = null
+	for (var/atom/A in targets)
 		if(A == src)
 			continue
+		var/datum/callback/cb = null
+		for (var/type in target_type_validator_map)
+			if (istype(A, type))
+				cb = target_type_validator_map[type]
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: Needs a break here.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 10]
```diff
@@ -300,3 +290,32 @@ mob/living/simple_animal/hostile/hitby(atom/movable/AM as mob|obj,var/speed = TH
 		spawn(10)
 			if(!src.stat)
 				horde()
+
+//////////////////////////////
+///////VALIDATOR PROCS////////
+//////////////////////////////
+
+/mob/living/simple_animal/hostile/proc/living(var/mob/living/L)
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: Could use better naming here. Eg "validator_living".

File: code/modules/mob/living/simple_animal/hostile/giant_spider.dm [hunk 11]
```diff
@@ -106,6 +108,16 @@
...
 	walk(src, 0)
 
+/mob/living/simple_animal/hostile/giant_spider/proc/validator_e_field(var/obj/effect/energy_field/E)
+	if(isliving(T)) // We prefer mobs over anything else
+		return FALSE
+	var/dist = get_dist(src, E)
+	if(dist < closest_distance)
+		closest_distance = dist
+		return TRUE
+	else
```
Comment (BurgerLUA) [inline code/modules/mob/living/simple_animal/hostile/giant_spider.dm]: Nitpick, you don't need this else statement at the end. Just return false if it fails. This is present throughout the PR.

Comment (Sindorman) [reply code/modules/mob/living/simple_animal/hostile/giant_spider.dm]: This false is required as it will be called several times on multiple shields, if we have 2 or more shields in our targets list, we want to target closest, and the way FindTarget() loops over this requires to return false, otherwise we have mobs targeting whichever shield was first in the list.

Comment (Sindorman) [reply code/modules/mob/living/simple_animal/hostile/giant_spider.dm]: That it’s not written in poor way, it is logically equivalent as there are only one costing that determines outcome.

Comment (skull132) [reply code/modules/mob/living/simple_animal/hostile/giant_spider.dm]: A stylistic choice. Also, arguably not an invalid one. A clear `if return A; else return B` indicates that both are of equal logical consequence. Where as a fall through would indicate that the `if return A;` is a special case.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 12]
```diff
@@ -18,8 +18,20 @@
...
+	var/list/target_type_validator_map = list()
+	var/lowest_health = INFINITY // Max you can get
+	var/atom/T = null
+
+/mob/living/simple_animal/hostile/Initialize()
+	. = ..()
+	target_type_validator_map[/mob/living] = CALLBACK(src, .proc/validator_living)
+	target_type_validator_map[/obj/mecha] = CALLBACK(src, .proc/validator_mecha)
+	target_type_validator_map[/obj/machinery/bot] = CALLBACK(src, .proc/validator_bot)
+	return .
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: okay style is one thing but `return .` is redundant once `.` has been initialized.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 13]
```diff
@@ -18,8 +18,20 @@
 	hunger_enabled = 0//Until automated eating mechanics are enabled, disable hunger for hostile mobs
 	var/shuttletarget = null
 	var/enroute = 0
+
+	// Vars to help find targets
 	var/list/targets = list()
 	var/attacked_times = 0
+	var/list/target_type_validator_map = list()
+	var/lowest_health = INFINITY // Max you can get
+	var/atom/T = null
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: This _has_ to be renamed. `T` is not descriptive, which is why variables named like that only exist within current scope, where their use can be quickly dicerned.

Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: Nit pick vol two. This is unnecessary. Expand the callback signature to include the current picked target, since it might be useful information. `validator_living(var/mob/living/L, var/atom/current)`.

I wonder if the same is worth doing for lowest health? Actually, that can be accessed if you cast current to mob and fetch it there!

File: code/modules/mob/living/simple_animal/hostile/carp.dm [hunk 14]
```diff
@@ -40,22 +40,91 @@
 	faction = "carp"
 
 	flying = TRUE
+	var/closest_distance = INFINITY
+
+/mob/living/simple_animal/hostile/carp/Initialize()
+	..()
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/carp.dm]: This has to be `. = ..()`

File: code/modules/mob/living/simple_animal/hostile/carp.dm [hunk 15]
```diff
@@ -40,22 +40,91 @@
...
 
 /mob/living/simple_animal/hostile/carp/Allow_Spacemove(var/check_drift = 0)
 	return 1	//No drifting in space for space carp!	//original comments do not steal
 
 /mob/living/simple_animal/hostile/carp/FindTarget()
 	. = ..()
-	if(.)
-		custom_emote(1,"nashes at [.]")
+	closest_distance = INFINITY
+	return .
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/carp.dm]: `return .` is a duplication. If `.` is assigned, it is returned by default.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 16]
```diff
@@ -18,8 +18,18 @@
...
 	var/enroute = 0
+
+	// Vars to help find targets
 	var/list/targets = list()
 	var/attacked_times = 0
+	var/list/target_type_validator_map = list()
+	var/lowest_health = INFINITY // Max you can get
+
+/mob/living/simple_animal/hostile/Initialize()
+	..()
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: Must be `. = ..()`

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 17]
```diff
@@ -300,3 +289,32 @@ mob/living/simple_animal/hostile/hitby(atom/movable/AM as mob|obj,var/speed = TH
...
+//////////////////////////////
+///////VALIDATOR PROCS////////
+//////////////////////////////
+
+/mob/living/simple_animal/hostile/proc/validator_living(var/mob/living/L, var/atom/current)
+	if((L.faction == src.faction) && !attack_same)
+		return FALSE
+	if(L in friends)
+		return FALSE
+	if(!L.stat && (L.health < lowest_health))
```
Comment (skull132) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: How to acquire lowest health:
```dm
	var/current_health = INFINITY
	if (isliving(current))
		var/mob/living/M = current
		current_health = M.health
```
